Add tests

ljvmiranda921 opened the pull request:
Resolves #14 

This requires a test `sqlite` database. It's  around 400 MB so I didn't decide to commit it. My suggestion is that everytime it performs a check, it downloads the `sqlite` file then execute `pytest`

## Things I need second opinion on:
- How I loaded the spatialite extension in `backend.cores.sqlite`. I used an event listener, not sure if that's the best solution.
- My test cases in `tests.spells.distance_to_nearest`. There should be a better way to do that, not quite sure yet.

I am also testing on BigQuery, but that can be triggered off via:

```s
$ pytest -m "not bqtest"
```

## How to run the tests

1. Ensure that you have the test `.sqlite` file in `tests/data/source.sqlite`. I'll upload this in GCS soon
2. Go to root directory of the project then run:

```s
$ pytest
```

"works on my machine..."

![Screenshot from 2019-03-11 21-20-15](https://user-images.githubusercontent.com/12949683/54126892-8b2d2500-4443-11e9-8551-ef2c65b505d5.png)


![coverage](https://user-images.githubusercontent.com/12949683/54127197-350cb180-4444-11e9-93ad-3f73fc600f87.png)


## Next steps (for next sprint)

- Write tests for helper functions in `backend.bq` 
- Refactor, if needed, tests in spell

ljvmiranda921 committed:
Add backend tests (#14)

This commit adds a set of tests for the backend module of geomancer.

Signed-off-by: Lester James V. Miranda <lj@thinkingmachin.es>

ljvmiranda921 committed:
Update docstrings for backend.base and spells.bq

ljvmiranda921 committed:
Add SQLiteCore

This commit adds a new DBCore for SQLite databases. It requires
the `mod_spatialite.so` or `libspatialite.so` extension in order
to work.

Signed-off-by: Lester James V. Miranda <lj@thinkingmachin.es>

ljvmiranda921 committed:
Add SQLiteConfig to settings

ljvmiranda921 committed:
Update dbcore generation in backend

This commit replaces if-else with a dictionary. However, it would
be better to do the same thing for the DBCore.load() method (the
problem is that they have different parameter names)

Signed-off-by: Lester James V. Miranda <lj@thinkingmachin.es>

ljvmiranda921 committed:
Add pytest-cov to dependencies

ljvmiranda921 committed:
Add tests for spells (#14)

ljvmiranda921 committed:
Update SQLite backend

When uploading to SQLite, we will mimic what BigQuery does, create
a column called __index_level_0__  so that all queries remain the
same.

Signed-off-by: Lester James V. Miranda <lj@thinkingmachin.es>

ljvmiranda921 committed:
Use `host` for DBCore initialization

Instead of passing custom **kwargs like `client` or `path`, we just
abstract it using the variable `host`. With that, it should be easier
to abstract our tests

Signed-off-by: Lester James V. Miranda <lj@thinkingmachin.es>

ljvmiranda921 committed:
Format with black

ljvmiranda921 committed:
Update EXCEPT literal_column in queries

Turns out that not all dialects think of EXCEPT the same way. Instead
of using that command, we'll just pass a list comprehension of columns.

Signed-off-by: Lester James V. Miranda <lj@thinkingmachin.es>

ljvmiranda921 committed:
Refactor backend module API

There's no need for a general backend module. get_tables and
get_engine is now part and coupled with the active DBCore

Signed-off-by: Lester James V. Miranda <lj@thinkingmachin.es>

ljvmiranda921 committed:
Update test documentation (#14)

ljvmiranda921 committed:
Fix spatialite loading via an event listener

Commit:
Use proper platform extension when loading spatialite

marksteve commented on geomancer/backend/cores/sqlite.py:
```diff
@@ -0,0 +1,75 @@
+        table_id = uuid.uuid4().hex
+        conn = sqlite3.connect(self.host)
+
+        # Ensure that spatialite extension is enabled
```
Is this required? Does it not work without this?

ljvmiranda921 replied:
Yeah, I will remove that part

marksteve commented on tests/backend/cores/abc_test_dbcore.py:
```diff
@@ -0,0 +1,52 @@
+from sqlalchemy.schema import Table
+
+
+class ABCTestDBCore(abc.ABC):
```
Why does it need to extend ABC?

ljvmiranda921 replied:
Hmmm, I thought that anyone who subclasses `ABCTestDBCore` should always implement the fixtures. I guess I can remove it na lang!

marksteve replied:
It's ok for APIs since it serves as a contract for usage but it's not so required in tests since it'll error out with the `NotImplementedError` exception anyways. Fixtures itself is already quite magical. This adds more learning overhead to contributors who don't know what ABC is.

marksteve commented on tests/backend/cores/abc_test_dbcore.py:
```diff
@@ -0,0 +1,52 @@
+        assert isinstance(engine, Engine)
+
+    @pytest.mark.usefixtures("core", "config", "sample_points", "test_tables")
+    def test_get_tables_return_type(
```
Not a useful test? Better checks IMO:
- Source table name is same same as input?
- Target table name is of uuid format?
- Columns are as expected?

ljvmiranda921 replied:
Ahh that makes sense. Sige will add those checks later!

marksteve commented on tests/spells/abc_test_spell.py:
```diff
@@ -0,0 +1,50 @@
+    """Base Test class for all spell implementations"""
+
+    @pytest.fixture
+    def spellhost(self):
```
This is cool 😎

ljvmiranda921 replied:
:sunglasses:

ljvmiranda921 committed:
Set LOGURU_LEVEL to INFO

Turns out that the default logging level for Loguru is DEBUG. We
don't want to show a lot of logging messages. Let's just set it to INFO

Signed-off-by: Lester James V. Miranda <lj@thinkingmachin.es>

ljvmiranda921 committed:
Add better verification tests for DBCore

New tests
- Check if source and target column names are as expected
- Check if source and target table names are as expected

Signed-off-by: Lester James V. Miranda <lj@thinkingmachin.es>

ljvmiranda921 committed:
Remove mod extension to SQLiteCore.load()

ljvmiranda921 commented:
Addressed your comments na, @marksteve
